examples/improved_wgan.py

```diff
@@ -0,0 +1,263 @@
+    this function at all points in the input space. The compromise used in the paper is to choose random points
+    on the lines between real and generated samples, and check the gradients at these points.
+
+    In order to evaluate the gradients, we must first run samples through the discriminator and evaluate the loss.
+    Then we get the gradients of the trainable discriminator weights, flatten the gradient tensors
+    and concatenate them into one 1D gradient tensor. The l2 norm and penalty can then be calculated from it.
+
+    Note that this loss function requires the model's trainable weights as input, but Keras only supports passing
+    y_true and y_pred to loss functions. To get around this, we make a partial() of the function with the
+    trainable_weights argument, and use that for model training."""
+    averaged_samples_loss = K.mean(y_pred)
+    gradients = K.gradients(averaged_samples_loss, trainable_weights)
+    gradients = K.concatenate([K.flatten(tensor) for tensor in gradients])
+    gradient_l2_norm = K.sqrt(K.sum(K.square(gradients)))
+    gradient_penalty = 2 * GRADIENT_PENALTY_WEIGHT * K.square(1 - gradient_l2_norm)
+    return gradient_penalty
+
+
+def make_generator():
+    """Creates a generator model that takes a 100-dimensional noise vector as a "seed", and outputs images
+    of size 28x28x1."""
+    model = Sequential()
+    model.add(Dense(1024, input_dim=100))
+    model.add(LeakyReLU())
+    model.add(Dense(128 * 7 * 7))
+    model.add(BatchNormalization())
+    model.add(LeakyReLU())
+    model.add(Reshape((7, 7, 128), input_shape=(128 * 7 * 7,)))
+    model.add(Conv2DTranspose(128, (5, 5), strides=2, padding='same'))
+    model.add(BatchNormalization())
+    model.add(LeakyReLU())
+    model.add(Convolution2D(64, (5, 5), padding='same'))
+    model.add(BatchNormalization())
+    model.add(LeakyReLU())
+    model.add(Conv2DTranspose(64, (5, 5), strides=2, padding='same'))
+    model.add(BatchNormalization())
+    model.add(LeakyReLU())
+    # Because we normalized training inputs to lie in the range [-1, 1],
+    # the tanh function should be used for the output of the generator to ensure its output
+    # also lies in this range.
+    model.add(Convolution2D(1, (5, 5), padding='same', activation='tanh'))
+    model.add(Reshape((28, 28, 1)))
+    return model
+
+
+def make_discriminator():
+    """Creates a discriminator model that takes an image as input and outputs a single value, representing whether
+    the input is real or generated. Unlike normal GANs, the output is not sigmoid and does not represent a probability!
+    Instead, the output should be as large and negative as possible for generated inputs and as large and positive
+    as possible for real inputs.
+
+    Note that the improved WGAN paper suggests that BatchNormalization should not be used in the discriminator."""
+    model = Sequential()
+    model.add(Convolution2D(64, (5, 5), padding='same', input_shape=(28, 28, 1)))

```

Comment:
Apologies, I have to stop assuming that everyone's working in Tensorflow style. I'll update that soon.